File: src/main/java/de/logicline/fruitbagger/CreateBagHandler.java
package de.logicline.fruitbagger;

import de.logicline.fruitbagger.domain.FruitBag;
import de.logicline.fruitbagger.domain.FruitUser;
import de.logicline.fruitbagger.domain.Session;
import io.vertx.core.Handler;
import io.vertx.ext.web.RoutingContext;
import org.mongodb.morphia.Datastore;

import java.util.List;
import java.util.OptionalInt;

public class CreateBagHandler implements Handler<RoutingContext> {
  private final Datastore datastore;

  public CreateBagHandler(Datastore datastore) {
    this.datastore = datastore;
  }

  @Override
  public void handle(RoutingContext ctx) {
    String sessionId = ctx.request().getParam("sessionId");
    FruitUser fruitUser = ctx.session().get("fruitUser");

    List<Session> fruitSessions = datastore.find(Session.class).field("user").equal(fruitUser).field("number").equal(Integer.valueOf(sessionId)).asList();

    if (fruitSessions.isEmpty()) {
      ctx.fail(new Exception("Session not found. Create a new one."));
      return;
    }

    Session fruitSession = fruitSessions.get(0);

    if(fruitSession.getFinishDate() != null){
      ctx.fail(new Exception("Session " + sessionId + " already closed!"));
      return;
    }

    // any open bags?
    List<FruitBag> openBags = datastore.find(FruitBag.class).field("session").equal(fruitSession).field("finishDate").equal(null).asList();
    if (!openBags.isEmpty()) {
      ctx.fail(new Exception("You have unfinished fruit bags left. Don't try to fool me! Your open bag: " + openBags.get(0).getNumber()));
      return;
    }


    OptionalInt max = datastore.find(FruitBag.class).field("session").equal(fruitSession).asList().stream()
      .mapToInt(s -> s.getNumber() == null ? 0 : s.getNumber())
      .max();
    FruitBag newBag = FruitBag.create(fruitSession, max.orElse(0) + 1);
    datastore.save(newBag);
    ctx.response().end(newBag.getNumber().toString());
  }
}

File: src/main/java/de/logicline/fruitbagger/CreateBagHandler.kt
package de.logicline.fruitbagger

import de.logicline.fruitbagger.domain.FruitBag
import de.logicline.fruitbagger.domain.FruitUser
import de.logicline.fruitbagger.domain.Session
import io.vertx.core.Handler
import io.vertx.ext.web.RoutingContext
import org.mongodb.morphia.Datastore

class CreateBagHandler(private val datastore: Datastore) : Handler<RoutingContext> {

    override fun handle(ctx: RoutingContext) {
        val sessionId = ctx.request()
            .getParam("sessionId")
        val fruitUser = ctx.session()
            .get<FruitUser>("fruitUser")

        val fruitSessions = datastore.find(Session::class.java)
            .field("user")
            .equal(fruitUser)
            .field("number")
            .equal(Integer.valueOf(sessionId))
            .asList()
        if (fruitSessions.isEmpty()) {
            ctx.fail(Exception("Session not found. Create a new one."))
            return
        }

        val fruitSession = fruitSessions[0]
        if (fruitSession.finishDate != null) {
            ctx.fail(Exception("Session $sessionId already closed!"))
            return
        }

        // get all bags
        val allBags = datastore.find(FruitBag::class.java)
            .field("session")
            .equal(fruitSession)
            .asList()

        // filter bags that are note finished
        var emptyBags = allBags.filter { bag -> bag.finishDate == null }
        if (!emptyBags.isEmpty()) {
            ctx.fail(Exception("You have unfinished fruit bags left. Don't try to fool me! Your open bag: " + allBags[0].number!!))
            return
        }


        // calculate the new id
        val max = allBags.stream()
            .mapToInt { s -> if (s.number == null) 0 else s.number }
            .max()
        val bagCount = max.orElse(0) + 1
        val newBag = FruitBag.create(fruitSession, bagCount)
        fruitSession.bagCount = bagCount
        datastore.save(newBag)
        datastore.save(fruitSession)
        ctx.response()
            .end(newBag.number!!.toString())
    }
}
